docs/smtchecker.rst

```diff
@@ -0,0 +1,733 @@
+|implementation                     |result is nondeterminisc.             |
+|                                   |CHC: Nondeterministic summary.        |
+|                                   |Try to infer invariants that hold     |
+|                                   |after the call returns.               |
++-----------------------------------+--------------------------------------+
+|transfer                           |BMC: Checks whether the contract's    |
+|                                   |balance is sufficient.                |
+|                                   |CHC: does not yet perform the check.  |
++-----------------------------------+--------------------------------------+
+|others                             |Currently unsupported                 |
++-----------------------------------+--------------------------------------+
+
+Using abstraction means loss of precise knowledge, but in many cases it does
+not mean loss of proving power.
+
+::
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >=0.8.0;
+    pragma experimental SMTChecker;
+    // This may report a warning if no SMT solver available.
+
+    contract Recover
+    {
+        function f(
+            bytes32 hash,
+            uint8 _v1, uint8 _v2,
+            bytes32 _r1, bytes32 _r2,
+            bytes32 _s1, bytes32 _s2
+        ) public pure returns (address) {
+            address a1 = ecrecover(hash, _v1, _r1, _s1);
+            require(_v1 == _v2);
+            require(_r1 == _r2);
+            require(_s1 == _s2);
+            address a2 = ecrecover(hash, _v2, _r2, _s2);
+            assert(a1 == a2);
+            return a1;
+        }
+    }
+
+In the example above, the SMTChecker is not expressive enough to actually
+compute ``ecrecover``, but by modelling the function calls as uninterpreted
+functions we know that the return value is the same when called on equivalent
+parameters. This is enough to prove that the assertion above is always true.
+
+Abstracting a function call with an UF can be done for functions known to be
+deterministic, and can be easily done for pure functions.  It is however
+difficult to do this with general external functions, since they might depend
+on state variables.
+
+Reference Types and Aliasing
+============================
+
+Solidity implements aliasing for reference types with the same :ref:`data
+location<data-location>`.
+That means one variable may be modified through a reference to the same data
+area.
+The SMTChecker does not keep track of which references refer to the same data.
+This implies that whenever a local reference or state variable of reference
+type is assigned, all knowledge regarding variables of the same type and data
+location is erased.
+If the type is nested, the knowledge removal also includes all the prefix base
+types.
+
+::
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >=0.8.0;
+    pragma abicoder v2;
```

Comment:
Does example depend on this? Also it is default n 0.8.0

Reply:
It depends on v2 because of `uint[][]` parameter in a public function. But true it's default now, so will remove.